### PR TITLE
New round notification, leader timeout value

### PR DESCRIPTION
**Pull request by afck**
## Motivation

For the BFT consensus protocol, validators will need to notify clients when a new consensus round starts, and sign to skip a round if it times out.

## Proposal

Add a `NewRound` notification reason, and a `LeaderTimeout` certificate value.

## Test Plan

These values aren't used yet, so there is nothing to test. This PR is in preparation for the BFT consensus, to keep the diff with the main branch small.

## Links

Public chains issue: https://github.com/linera-io/linera-protocol/issues/464
BFT consensus draft: https://github.com/linera-io/linera-protocol/pull/928

## Release Plan

- [x] All good!
- [ ] Need to bump the major/minor version number in the next release of the crates.
- [ ] Need to update the developer manual.
- [ ] This PR is adding or removing Cargo features.
- [ ] Release is blocked and/or tracked by other issues (see links above)

## Reviewer Checklist

- [ ] The title and the summary of the PR are short and descriptive.
- [ ] The proposed solution achieves the goals stated in the PR.
- [ ] The test plan is reproducible and provides sufficient coverage.
- [ ] The release plan is adequate.
- [ ] The commits correspond to distinct logical changes.
- [ ] The code follows the [coding guidelines](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md).
- [ ] The proposed changes look correct.
- [ ] The CI is passing.
- [x] All of the above!

### CODE REVIEW

File: linera-core/src/client.rs [hunk 1]
```diff
@@ -368,18 +368,27 @@ where
         Ok(self.key_pair().await?.public())
     }
 
-    async fn get_local_next_block_height(
+    async fn get_local_chain_info(
```
Comment (ma2bd) [inline linera-core/src/client.rs:371]: nit: Drop `get_` for naming consistency?

File: linera-core/src/client.rs [hunk 2]
```diff
@@ -460,6 +469,34 @@ where
                     }
                 }
             }
+            Reason::NewRound { height, round } => {
+                let chain_id = notification.chain_id;
+                if let Some(info) =
+                    Self::get_local_chain_info(this.clone(), chain_id, &mut local_node).await
+                {
+                    if info.next_block_height > height || info.manager.next_round() > round {
```
Comment (ma2bd) [inline linera-core/src/client.rs:477]: This could skip `(info.next_block_height + 1, 0)`. Don't we want `(info.next_block_height, info.manager.next_round()) > (height, round)` here?

Actually, is it a good idea to separate `NewRound` from `NewBlock`? I don't think we guarantee that notifications arrive in order.

**Review by ma2bd — COMMENTED** [review]
LGTM with one question about round notifications though

File: linera-core/src/client.rs (hunk 2 above)
Comment (afck) [reply linera-core/src/client.rs:477]: Ah, good catch! :facepalm: 

Yes, we could remove `NewBlock` now—it's implied by a `NewRound` for round 0.

But many clients will only be interested in new blocks, not new rounds, so it could be convenient to keep them separate. It shouldn't hurt that they don't arrive in order?

**Review by maxtori — APPROVED** [review]

File: linera-core/src/client.rs (hunk 2 above)
Comment (ma2bd) [reply linera-core/src/client.rs]: It creates a bit of duplication for those clients who actually care about rounds but ok.

File: linera-core/src/worker.rs [hunk 3]
```diff
@@ -638,33 +642,82 @@ where
...
         }
         if self
             .cache_recent_value(Cow::Borrowed(&certificate.value))
             .await
         {
-            let value = certificate.value.clone().into_confirmed();
-            self.cache_recent_value(Cow::Owned(value)).await;
+            if let Some(value) = certificate.value.clone().into_confirmed() {
+                self.cache_recent_value(Cow::Owned(value)).await;
+            }
```
Comment (ma2bd) [inline linera-core/src/worker.rs:662]: Seems like we could have a small helper function taking `&mut self` and a `&Value`.

File: linera-core/src/worker.rs [hunk 4]
```diff
@@ -801,29 +836,29 @@ where
...
+                let application_id = outgoing_message.message.application_id();
+                Medium::Channel(ChannelFullName {
                     application_id,
                     name,
-                }),
-            },
+                })
+            }
         };
+        let origin = Origin { sender, medium };
```
Comment (ma2bd) [inline linera-core/src/worker.rs:861]: It's a bit hard to parse the lines but this seems like a vanilla code refactoring, right? Consider a separate commit next time to speed up reviewing.

**Review by ma2bd — APPROVED** [review]
Thanks!

File: linera-core/src/worker.rs (hunk 4 above)
Comment (afck) [reply linera-core/src/worker.rs:856]: Right, sorry.
It's mostly just refactoring, but also dealing with `messages` now returning an option.
It got rebased, squashed and split up a few times already, and I failed to properly separate the pure refactoring part in the end. :sweat_smile:
